Menus/UndoMenu/UndoMenu.cs

```diff
@@ -0,0 +1,204 @@
+#if UNITY_EDITOR
+using System;
+using System.Collections;
+using System.Collections.Generic;
+using UnityEditor.Experimental.EditorVR.Core;
+using UnityEditor.Experimental.EditorVR.Extensions;
+using UnityEditor.Experimental.EditorVR.Proxies;
+using UnityEditor.Experimental.EditorVR.Utilities;
+using UnityEngine;
+using UnityEngine.InputNew;
+
+namespace UnityEditor.Experimental.EditorVR.Menus
+{
+    sealed class UndoMenu : MonoBehaviour, IInstantiateUI, IAlternateMenu, IUsesMenuOrigins, ICustomActionMap,
+        IControlHaptics, IUsesNode, IConnectInterfaces, IRequestFeedback, IUsesDeviceType
+    {
+        const float k_UndoRedoThreshold = 0.5f;
+        const float k_EngageUndoAfterStickReleasedDuration = 0.1f; // Duration after releasing the joystick to still accept a left/right flick to undo/redo.
+        const string k_FeedbackHintForJoystickController = "Click + flick left/right to undo/redo";
+        const string k_FeedbackHintForTrackpadController = "Click left/right side to undo/redo";
+
+        [SerializeField]
+        ActionMap m_ActionMap;
+
+        [SerializeField]
+        UndoMenuUI m_UndoMenuPrefab;
+
+        [SerializeField]
+        HapticPulse m_UndoPulse;
+
+        UndoMenuUI m_UndoMenuUI;
+        List<ActionMenuData> m_MenuActions;
+        Transform m_AlternateMenuOrigin;
+        MenuHideFlags m_MenuHideFlags = MenuHideFlags.Hidden;
+        float m_PrevNavigateX;
+        bool m_StillEngagedAfterStickRelease;
+        Coroutine m_StillEngagedAfterStickReleasedCoroutine;
+        bool m_TrackpadController;
+        string m_FeedbackHintForCurrentController;
+
+        readonly BindingDictionary m_Controls = new BindingDictionary();
+        
+        public Transform rayOrigin { private get; set; }
+
+        public Transform menuOrigin { get; set; }
+
+        public GameObject menuContent { get { return m_UndoMenuUI.gameObject; } }
+
+        public Node node { get; set; }
+
+        public event Action<Transform> itemWasSelected;
```

Comment:
There's a warning about this being unused. Does UndoMenu need to be an IAlternateMenu? Could it just be an IMenu? I think we can still just attach it to the alternate menu origin but make it not actually be a IAlternateMenu. Same goes with IUsesRayOrigin (Included in IAlternateMenu). We don't seem to use/need it.

It looks like this will require a refactor of SpawnAlternateMenu, etc. but to look at it, all we do is check that it's the right interface, add a component, and connect interfaces. I think we could make it generic no problem.